File: app/src/main/java/com/etesync/syncadapter/ui/setup/BaseConfigurationFinder.java
/*
 * Copyright © 2013 – 2015 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */
package com.etesync.syncadapter.ui.setup;

import android.content.Context;
import android.support.annotation.NonNull;

import com.etesync.syncadapter.HttpClient;
import com.etesync.syncadapter.journalmanager.Crypto;
import com.etesync.syncadapter.journalmanager.Exceptions;
import com.etesync.syncadapter.journalmanager.JournalAuthenticator;
import com.etesync.syncadapter.log.StringHandler;
import com.etesync.syncadapter.model.CollectionInfo;

import java.io.IOException;
import java.io.Serializable;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

import lombok.Getter;
import lombok.ToString;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

public class BaseConfigurationFinder {
    protected final Context context;
    protected final LoginCredentials credentials;

    protected final Logger log;
    protected final StringHandler logBuffer = new StringHandler();
    protected OkHttpClient httpClient;

    public BaseConfigurationFinder(@NonNull Context context, @NonNull LoginCredentials credentials) {
		this.context = context;
        this.credentials = credentials;

        log = Logger.getLogger("syncadapter.BaseConfigurationFinder");
        log.setLevel(Level.FINEST);
        log.addHandler(logBuffer);

        httpClient = HttpClient.create(context, log);
	}


    public Configuration findInitialConfiguration() {
        boolean failed = false;
        Configuration.ServiceInfo
                cardDavConfig = findInitialConfiguration(CollectionInfo.Type.ADDRESS_BOOK),
                calDavConfig = findInitialConfiguration(CollectionInfo.Type.CALENDAR);

        JournalAuthenticator authenticator = new JournalAuthenticator(httpClient, HttpUrl.get(credentials.uri));

        String authtoken = null;
        try {
            authtoken = authenticator.getAuthToken(credentials.userName, credentials.password);
        } catch (Exceptions.HttpException|IOException e) {
            log.warning(e.getMessage());

            failed = true;
        }

        return new Configuration(
                credentials.uri,
                credentials.userName, authtoken,
                cardDavConfig, calDavConfig,
                logBuffer.toString(), failed
        );
    }

    protected Configuration.ServiceInfo findInitialConfiguration(@NonNull CollectionInfo.Type service) {
        // put discovered information here
        final Configuration.ServiceInfo config = new Configuration.ServiceInfo();
        log.info("Finding initial " + service.toString() + " service configuration");

        return config;
    }

    // data classes

    @ToString(exclude={"logs", "authtoken", "rawPassword", "password"})
    public static class Configuration implements Serializable {
        // We have to use URI here because HttpUrl is not serializable!

        public Configuration(final URI url, final String userName, final String authtoken, final ServiceInfo cardDAV, final ServiceInfo calDAV, final String logs, final boolean failed) {
            this.url = url;
            this.userName = userName;
            this.authtoken = authtoken;
            this.cardDAV = cardDAV;
            this.calDAV = calDAV;
            this.logs = logs;
            this.failed = failed;
        }

        @ToString
        public static class ServiceInfo implements Serializable {
            public final Map<String, CollectionInfo> collections = new HashMap<>();
        }

        public final URI url;

        public final String userName, authtoken;
        public String rawPassword;
        public String password;
        public Crypto.AsymmetricKeyPair keyPair;

        public final ServiceInfo cardDAV;
        public final ServiceInfo calDAV;

        public final String logs;

        public Throwable error;

        @Getter
        private final boolean failed;
    }

}

File: app/src/main/java/com/etesync/syncadapter/ui/setup/BaseConfigurationFinder.kt
/*
 * Copyright © 2013 – 2015 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */
package com.etesync.syncadapter.ui.setup

import android.content.Context
import com.etebase.client.Account
import com.etebase.client.Client
import com.etebase.client.exceptions.EtebaseException
import com.etesync.syncadapter.HttpClient
import com.etesync.journalmanager.Crypto
import com.etesync.journalmanager.Exceptions
import com.etesync.journalmanager.JournalAuthenticator
import com.etesync.journalmanager.UserInfoManager
import com.etesync.syncadapter.Constants
import com.etesync.syncadapter.log.Logger
import com.etesync.syncadapter.model.CollectionInfo
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.io.Serializable
import java.net.URI
import java.util.*

class BaseConfigurationFinder(protected val context: Context, protected val credentials: LoginCredentials) {
    protected var httpClient: OkHttpClient

    init {
        httpClient = HttpClient.Builder(context).build().okHttpClient
    }

    private fun isServerEtebase(): Boolean {
        if (credentials.uri != null) {
            val remote = credentials.uri.toHttpUrlOrNull()!!.newBuilder()
                    .addPathSegments("api/v1/authentication/is_etebase/")
                    .build()

            val request = Request.Builder()
                    .get()
                    .url(remote)
                    .build()

            val response = httpClient.newCall(request).execute()

            return response.isSuccessful
        } else {
            return !credentials.userName.contains("@")
        }
    }

    fun findInitialConfigurationLegacy(): Configuration {
        var exception: Throwable? = null

        val uri = credentials.uri ?: URI(Constants.serviceUrl.toString())

        val authenticator = JournalAuthenticator(httpClient, uri.toHttpUrlOrNull()!!)

        var authtoken: String? = null
        var userInfo: UserInfoManager.UserInfo? = null
        try {
            authtoken = authenticator.getAuthToken(credentials.userName, credentials.password)

            val authenticatedHttpClient = HttpClient.Builder(context, uri.host, authtoken!!).build().okHttpClient
            val userInfoManager = UserInfoManager(authenticatedHttpClient, uri.toHttpUrlOrNull()!!)
            userInfo = userInfoManager.fetch(credentials.userName)
        } catch (e: Exceptions.HttpException) {
            Logger.log.warning(e.message)
            exception = e
        } catch (e: IOException) {
            Logger.log.warning(e.message)
            exception = e
        }

        return Configuration(
                uri,
                credentials.userName,
                null,
                authtoken,
                userInfo,
                exception
        )
    }

    fun findInitialConfigurationEtebase(): Configuration {
        var exception: Throwable? = null

        val uri = credentials.uri ?: URI(Constants.etebaseServiceUrl)

        var etebaseSession: String? = null
        try {
            val client = Client.create(httpClient, uri.toString())
            val etebase = Account.login(client, credentials.userName, credentials.password)
            etebaseSession = etebase.save(null)
        } catch (e: EtebaseException) {
            exception = e
        }

        return Configuration(
                uri,
                credentials.userName,
                etebaseSession,
                null,
                null,
                exception
        )
    }

    fun findInitialConfiguration(): Configuration {
        try {
            if (isServerEtebase()) {
                return findInitialConfigurationEtebase()
            } else {
                return findInitialConfigurationLegacy()
            }
        } catch (e: Exception) {
            return Configuration(
                    credentials.uri,
                    credentials.userName,
                    null,
                    null,
                    null,
                    e
            )
        }
    }

    // data classes

    class Configuration
    // We have to use URI here because HttpUrl is not serializable!

    (val url: URI?, val userName: String, val etebaseSession: String?, val authtoken: String?, var userInfo: UserInfoManager.UserInfo?, var error: Throwable?) : Serializable {
        var rawPassword: String? = null
        var password: String? = null
        var keyPair: Crypto.AsymmetricKeyPair? = null

        val isFailed: Boolean
            get() = this.error != null

        val isLegacy: Boolean
            get() = this.authtoken != null

        class ServiceInfo : Serializable {
            val collections: Map<String, CollectionInfo> = HashMap()

            override fun toString(): String {
                return "BaseConfigurationFinder.Configuration.ServiceInfo(collections=" + this.collections + ")"
            }
        }

        override fun toString(): String {
            return "BaseConfigurationFinder.Configuration(url=" + this.url + ", userName=" + this.userName + ", keyPair=" + this.keyPair + ", error=" + this.error + ", failed=" + this.isFailed + ")"
        }
    }

}
